File: BrianLee/Week4/2022-10-04-Coin-Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if(amount == 0) return 0;

        Map<Integer, Integer> mem = new HashMap<>();
        getNumberOfCoins(coins, 0, amount, mem);
        return  mem.get(amount) == Integer.MAX_VALUE ? -1: mem.get(amount);
    }

    public int getNumberOfCoins(int[] coins, int level, int leftAmount, Map<Integer, Integer> mem) {
        if(leftAmount == 0) return 0;
        if(leftAmount < 0) return Integer.MAX_VALUE;

        if(mem.containsKey(leftAmount)) {
            return mem.get(leftAmount);
        }

        int min = Integer.MAX_VALUE;
        for(int coin: coins) {
            int cur = getNumberOfCoins(coins, level+1, leftAmount-coin, mem);
            if(cur != Integer.MAX_VALUE) {
                min = Math.min(min, cur + 1);
            }
        }
        mem.put(leftAmount, min);

        return min;
    }
}

File: Chris/2022-10-04-Coin Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        dp = [amount+1] * (amount+1)
        
        dp[0] = 0
        
        for i in range(1,amount+1):
            for c in coins:
                
                if c <= i:
                    dp[i] = min(dp[i], dp[i-c] + 1)
                
        
        return dp[amount] if dp[amount] != (amount+1) else -1